src/unix/ibmi.c

```diff
@@ -247,3 +312,164 @@ int uv_cpu_info(uv_cpu_info_t** cpu_infos, int* count) {
 
   return 0;
 }
+
+
+static int get_ibmi_physical_address(const char *line, char *phys_addr)
+{
+  LIND0500 rcvr;
+  /* rcvrlen is input parameter 2 to QDCRLIND */
+  unsigned int i, rcvrlen = sizeof(rcvr);
+  unsigned char format[8], line_name[10];
+
+  /* format is input parameter 3 to QDCRLIND */
+  iconv_a2e("LIND0500", format, sizeof(format));
+
+  /* line_name is input parameter 4 to QDCRLIND */
+  iconv_a2e(line, line_name, sizeof(line_name));
+
+  /* err is input parameter 5 to QDCRLIND */
+  errcode_s err;
+
+  /* qwcrssts_pointer is the 16-byte tagged system pointer to QDCRLIND */
+  ILEpointer __attribute__((aligned(16))) qdcrlind_pointer;
+
+  /* qwcrssts_argv is the array of argument pointers to QDCRLIND */
+  void *qdcrlind_argv[6];
+
+  /* Set the IBM i pointer to the QSYS/QDCRLIND *PGM object */
+  int rc = _RSLOBJ2(&qdcrlind_pointer, RSLOBJ_TS_PGM, "QDCRLIND", "QSYS");
+
+  if (rc != 0)
+    return rc;
+
+  /* initialize the QDCRLIND returned info structure */
+  memset(&rcvr, 0, sizeof(rcvr));
+
+  /* initialize the QDCRLIND error code structure */
+  memset(&err, 0, sizeof(err));
+  err.bytes_provided = sizeof(err);
+
+  /* initialize the array of argument pointers for the QDCRLIND API */
+  qdcrlind_argv[0] = &rcvr;
+  qdcrlind_argv[1] = &rcvrlen;
+  qdcrlind_argv[2] = &format;
+  qdcrlind_argv[3] = &line_name;
+  qdcrlind_argv[4] = &err;
+  qdcrlind_argv[5] = NULL;
+
+  /* Call the IBM i QDCRLIND API from PASE */
+  rc = _PGMCALL(&qdcrlind_pointer, (void **)&qdcrlind_argv, 0);
+  if (rc != 0)
+    return rc;
+
+  unsigned char mac_addr[sizeof(rcvr.loca_adapter_address)];
+  char buf[2];
+
+  /* convert ebcdic loca_adapter_address to ascii first */
+  iconv_e2a(rcvr.loca_adapter_address, mac_addr, 
+            sizeof(rcvr.loca_adapter_address));
+
+  /* convert loca_adapter_address(char[12]) to phys_addr(char[6]) */
+  for (i = 0; i < 6; i++) {
+    char* endptr = NULL;
+    sprintf(buf, "%c%c", mac_addr[i * 2], mac_addr[i * 2 + 1]);
+    long n = strtol(buf, &endptr, 16);
+    if(endptr == NULL) {
+      /* invalid hex string */
+      *(phys_addr + i) = 0;
+      rc = -1;
+    }
+    else
+      *(phys_addr + i) = (unsigned char)n;
+  }
```

Comment:
This seems like a very convoluted way of parsing a fixed-size hexadecimal string. What about this?
```c
int c[6];
int rc;

rc = sscanf(mac_addr, "%02x%02x%02x%02x%02x%02x",
            &c[0], &c[1], &c[2], &c[3], &c[4], &c[5]);

if (rc == ARRAY_SIZE(c)) {
  phys_addr[0] = c[0];
  phys_addr[1] = c[1];
  phys_addr[2] = c[2];
  phys_addr[3] = c[3];
  phys_addr[4] = c[4];
  phys_addr[5] = c[5];
} else {
  /* handle error */
}
```

Reply:
Thanks for the recommendation. This looks much better.

Reply:
@bnoordhuis As for the error handler, how about this?
```
  int r = sscanf(mac_addr, "%02x%02x%02x%02x%02x%02x",
              &c[0], &c[1], &c[2], &c[3], &c[4], &c[5]);

  if (r == ARRAY_SIZE(c)) {
    phys_addr[0] = c[0];
    phys_addr[1] = c[1];
    phys_addr[2] = c[2];
    phys_addr[3] = c[3];
    phys_addr[4] = c[4];
    phys_addr[5] = c[5];
  } else {
    memset(phys_addr, 0, 6);
    rc = -1;
  }
```